File: python/second/max_path_sum.py
class Solution(object):
    # O(n) time | O(h) space
    def maxPathSum(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        res = [root.val]

        def dfs(node):
            if not node:
                return 0
            leftBranch = max(dfs(node.left), 0)
            rightBranch = max(dfs(node.right), 0)

            maxBranch = max(leftBranch, rightBranch)

            res[0] = max(res[0], node.val)
            res[0] = max(res[0], node.val + maxBranch)
            res[0] = max(res[0], node.val + leftBranch + rightBranch)

            return node.val + maxBranch

        dfs(root)
        return res[0]

        
class TreeNode(object):
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __repr__(self):
        result = str(self.val)
        result += str(self.left) if self.left else ''
        result += str(self.right) if self.right else ''
        return result


root = TreeNode(-10)
root.left = TreeNode(9)
root.right = TreeNode(20)
root.right.left = TreeNode(15)
root.right.right = TreeNode(7)

solution = Solution()
print(solution.maxPathSum(root))    

File: src/main/java/june_2025/MaxPathSum.java
package june_2025;

public class MaxPathSum {

    public static void main(String[] args) {
//        TreeNode root = new TreeNode(-10);
//        root.left = new TreeNode(9);
//        root.right = new TreeNode(20);
//        root.right.left = new TreeNode(15);
//        root.right.right = new TreeNode(7);

        TreeNode root = new TreeNode(-10);
//        root.left = new TreeNode(2);
//        root.right = new TreeNode(3);

        MaxPathSum maxPathSum = new MaxPathSum();
        int result = maxPathSum.maxPathSum(root);
        System.out.println(result);

    }

    int maxPath;

    public int maxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        maxPath = root.val;
        dfs(root);
        return maxPath;
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int left = Math.max(dfs(node.left), 0);
        int right = Math.max(dfs(node.right), 0);
        int maxBranch = Math.max(left, right);
        maxPath = Math.max(maxPath, node.val);
        maxPath = Math.max(maxPath, node.val + maxBranch);
        maxPath = Math.max(maxPath, node.val + left + right);
        return maxBranch + node.val;
    }

    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        public TreeNode(int val) {
            this.val = val;
        }
    }

}
